### PR TITLE
Add a non-recursive mutex.

### PR DESCRIPTION
**Pull request by ombre5733**
As recursive mutexes are frequently considered to be less secure compared to plain mutexes (see [here](https://cppwisdom.quora.com/Recursive-mutexes-considered-harmful) or [here](http://www.zaval.org/resources/library/butenhof1.html), for example), let's add a non-recursive variant to the RTOS.

The change is quite straightforward: An unused padding byte of the mutex structure is used for storing a flag which discriminates plain and recursive mutexes. The flag is checked in `rt_mut_wait()`. If a task tries to re-lock a plain mutex which it has already acquired, the nesting level is not changed and an error is returned.

In order not to break the existing API, a function `osPlainMutexCreate()` has been added to
create a plain (non-recursive) mutex. The existing function `osMutexCreate()` is used to initialize a recursive mutex as usual.

Alternatively, one could change the signature of

```
osMutexId osMutexCreate (const osMutexDef_t *mutex_def);
```

to

```
osMutexId osMutexCreate (const osMutexDef_t *mutex_def, os_mutex_type type);
```

and pass the `type` of the mutex as parameter. The latter is more elegant in my opinion but it breaks the existing API, of course.

### CODE REVIEW

**Comment by ReinhardKeil** [comment]
Thanks for this feedback.  We have looked into it, but before we drive it forward we would like to get more information.

Actually we think that similar functionality could be achieved with counting semaphores (yes I know we need to improve the handling of that).  We would therefore like to understand the use-case of a plain mutex compared to a counting semaphore.

**Comment by ombre5733** [comment]
Unfortunately, a semaphore won't do in this case. There are at least two drawbacks using it compared to a mutex:
1. There is no guarantee that the thread which acquires a semaphore token also releases it. So in theory, it could be that thread A acquires a token, thread B releases a token and thread C both acquires and releases the token. This makes reasoning about the critical section quite difficult.
2. Semaphores do not prevent priority-inversion. Assume that a low-priority thread has acquired the last semaphore token and a high-priority thread blocks on the semaphore. If a mid-priority thread (which does not use the semaphore at all) prevents the low-priority thread from running, the high-priority thread will starve.

While the first argument above might sound artificial, using a semaphore to protect a criticial section does not feel right. Semaphores are made for thread-safe counting but mutexes are made to restrict access to a shared resource to a single thread. While it might be possible to abuse a semaphore as a mutex, it does definitly not provide the same level of quality as a mutex. And when it comes to advanced topics such as priority-inversion, an implementation based on semaphores may be simply wrong.

So the question is, why do we need non-recursive mutexes at all?

I would argue that probably 95% of all mutexes are not needed to be recursive. In our application base it happens quite rarely that a mutex must be locked twice by the same thread and even in cases where this happens, it can be argued if the code has an easy-to-understand structure (when a thread acquires a mutex multiple times, the synchronization points are often not clear and the code locks a mutex everywhere just to be sure).

Non-recursive mutex are great for detecting programming errors. Let me give a few examples.

Most often, I lock and unlock a mutex in the same function:

```
void a()
{
    osStatus ret;
    ret = osMutexWait(mutex_id, osWaitForever);
    if (ret != osOK) handle_error();

    // (*) At this point, the shared resource is in a consistent state.

    ... modify the shared resource ...

    ... bring the shared resource into a consistent state again ...

    osMutexRelease(mutex_id);
}
```

The mental model is that right after the mutex is locked and just before the mutex is released, the shared resource is in a consistent state. For example, if the resource contains an array, the counter keeping track of the number of elements in the array and the array data itself are both valid and in-sync; if it consists of more than one variable, certain relations or invariants between the variables hold...

Now consider that the function `a()` above is called from another function `b()`, which itself also locks the mutex.

```
void b()
{
    osStatus ret;
    osMutexWait(mutex_id, osWaitForever);
    if (ret != osOK) handle_error();

    ... modify the shared resource ...
    a();
    ... bring the shared resource into a consistent state again ...
    osMutexRelease(mutex_id);
}
```

This has introduced an error now: at the line marked `(*)` in function `a()`, the shared resource is in no consistent state any longer because `b()` has modified it before calling `a()`. Unfortunately, our mental model is wrong. After locking the mutex, the shared resource is in a consistent state **only if** the mutex was locked for the first time.

Let me give a more concrete example of what can go wrong with recursive mutexes. Suppose we have a data base stored in some flash memory. A thread can load one record into a shared buffer, modify it and store the record back to the flash:

```
osMutexId databaseMutex;
int recordId;
char recordBuffer[128];

int DB_openRecord(int id)
{
    // Lock the data base.
    osStatus ret = osMutexWait(databaseMutex, osWaitForever);
    if (ret != osOK)
        return -1;

    recordId = id;
    ... load data from storage into recordBuffer ...
    ... make sanity checks in recordBuffer (checksum etc) ...
}

int DB_modifyRecord(...)
{
    ... update recordBuffer ...
}

int DB_closeRecord(void)
{
    ... finalize recordBuffer (checksum etc) ...
    ... write recordBuffer back into data base storage ...

    // Release the data base.
    osMutexRelease(databaseMutex);
}
```

A mutex is locked when a record is loaded and it is released when the record is closed again. This seems fine until we made an error and loaded two records in the same thread forgetting to close the first one:

```
DB_openRecord(1);
... quite some piece of code ...
... forgot to close record 1 here ...

DB_openRecord(2);
... modifications to the record
DB_closeRecord();

... code which originally closed record 1 ...
DB_closeRecord();
```

Unfortunately, `DB_openRecord(2)` did not return an error because locking the mutex twice by the same thread is allowed. To catch these silly programming errors (which again took a lot of time to debug on embedded hardware), we have to add a flag keeping track if the data base has already been locked before, e.g.

```
int databaseLocked = 0; // If set, the data base has already been locked.

int DB_openRecord(int id)
{
    // Lock the data base.
    osStatus ret = osMutexWait(databaseMutex, osWaitForever);
    if (ret != osOK)
        return -1;
    if (databaseLocked)
        return -1;
    // Explicitly mark the data base as locked.
    databaseLocked = 1;

    recordId = id;
    ... load data from storage into recordBuffer ...
    ... make sanity checks in recordBuffer (checksum etc) ...
}
```

So `databaseLocked` makes the mutex non-recursive but this is now exactly the flag that I propose to add to the mutex to prevent recursive locking.

One last example, which demonstrates the usefullness of non-recursive mutexes is in the context of condition variables (CV). Now how does a CV work? One thread - the mutator - locks a mutex, modifies a shared state, unlocks the mutex and sends a CV-signal to another thread. The observer locks the mutex, tests if the shared state fulfills a certain condition and if not, blocks on the condition variable. Conceptually:

```
osMutexId mutex;
cond_var cv;
struct SharedState shared_state;

void mutator(void)
{
    osStatus ret;
    // Modify the shared state while the mutex is locked.
    ret = osMutexWait(mutex, osWaitForever);
    if (ret != osOK) handle_error();

    ... modify shared_state ...

    osMutexRelease(mutex);

    // Notify the observer about the changes to shared_state.
    cv_notify_one(&cv);
}

void observer(void)
{
    osStatus ret;
    ret = osMutexWait(mutex, osWaitForever);
    if (ret != osOK) handle_error();

    while (shared_state does not fulfill condition)
    {
        // Release the mutex, block until we reach a notification via
        // the condition variable and lock the mutex again.
        cv_wait(&cv, mutex);
    }

    ... do something with shared_state ...
    osMutexRelease(mutex);
}
```

The function `cv_wait()` will unlock the mutex, block until it reaches a signal via the given condition variable and re-lock the mutex again. During the time when the mutex is unlocked, the mutator can modify the shared state and send a notification to the observer.

Now what happens if an observer locks the mutex twice?

```
void evil_observer(void)
{
    osStatus ret;
    ret = osMutexWait(mutex, osWaitForever);
    if (ret != osOK) handle_error();

    ... some code ...

    ret = osMutexWait(mutex, osWaitForever);
    if (ret != osOK) handle_error();

    while (shared_state does not fulfill condition)
    {
        // Release the mutex, block until we reach a notification via
        // the condition variable and lock the mutex again.
        cv_wait(&cv, mutex);
    }

    ... do something with shared_state ...
    osMutexRelease(mutex);
    ... some code ...
    osMutexRelease(mutex);
}
```

Well, this code leads to a dead-lock. The reason is that `cv_wait()` will unlock the mutex **once** and wait until a signal is received via the condition variable. Unfortunately, the mutator will never send a signal because it can only do so when it has acquired the mutex which is still held by the thread which called `evil_observer()`. So in this case the usage of a recursive mutex leads to a dead-lock, which is again very hard to debug. Note that in the example above the double-locking is easily detected as `osMutexWait()` is called twice in the same function but this is not the case when the locking is split between two different functions as in the first example.

All the problems above are easily avoided if the mutex was not recursive, i.e. if calling `osMutexWait()` returned an error when called twice from the same thread.

So in summary, I think that a plain mutex adds a lot of value to the OS due to the fact that it is not possible to **accidentally** lock it multiple times, i.e. locking it twice fails visibly.

**Comment by ombre5733** [comment]
Ping. Any comments on my use-cases?
